### PR TITLE
separate planning strategies from DevicesPlanner into DevicesPlannerStrategies module

### PR DESCRIPTION
**Pull request by wfeldt**
There are currently `:legacy` and a stub for `:ng`.

The `:legacy` code is basically a 1:1 copy of the old `devices_planner.rb`.

### CODE REVIEW

File: src/lib/y2storage/proposal/devices_planner_strategies/legacy.rb [hunk 1]
```diff
@@ -0,0 +1,253 @@
+#!/usr/bin/env ruby
```
Comment (joseivanlopez) [inline src/lib/y2storage/proposal/devices_planner_strategies/legacy.rb]: The class should not be considered a script itself, so we have not to include a shebang.

File: src/lib/y2storage/proposal/devices_planner_strategies/ng.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/env ruby
```
Comment (joseivanlopez) [inline src/lib/y2storage/proposal/devices_planner_strategies/ng.rb]: Same here.

File: src/lib/y2storage/proposal/devices_planner_strategies/ng.rb [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+#
+# encoding: utf-8
+
+# Copyright (c) [2016-2017] SUSE LLC
```
Comment (joseivanlopez) [inline src/lib/y2storage/proposal/devices_planner_strategies/ng.rb]: Strictly speaking, this should be only 2017.

File: src/lib/y2storage/proposal/devices_planner.rb [hunk 4]
```diff
@@ -35,215 +31,33 @@ module Proposal
...
-      MAX_SWAP_SIZE = DiskSize.GiB(2) # This is also DEFAULT_SWAP_SIZE
-      SWAP_WEIGHT = 100 # Importance of swap if low free disk space
+      attr_accessor :settings
 
       def initialize(settings, devicegraph)
         @settings = settings
         @devicegraph = devicegraph
+        strategy = @settings.format
+        @strategy = STRATEGIES[strategy].new(settings, devicegraph)
+      rescue
```
Comment (joseivanlopez) [inline src/lib/y2storage/proposal/devices_planner.rb]: Here, an exception could be raised for different reasons (unknown strategy, failure in new, etc). Maybe, we should check specifically the correctness of the strategy:

```
def initialize(settings, devicegraph)
  @settings = settings
  @devicegraph = devicegraph
  @strategy = strategy
end

def strategy
  return @strategy unless @strategy.nil?

  strategy_class = 
    case @settings.format
    when :legacy
      DevicesPlannerStrategies::Legacy
    when :ng
      DevicesPlannerStrategies::Ng
    else
      err_msg = "Unsupported device planner strategy :#{@settings.format}"
      log.error err_msg
      raise DevicesPlannedStrategyError, err_msg
    end

  @strategy = strategy_class.new(@settings, @devicegraph)
end
```

Comment (joseivanlopez) [reply src/lib/y2storage/proposal/devices_planner.rb]: In this case, the constant `DevicesPlanner::STRATEGIES` would not be necessary.

File: src/lib/y2storage/proposal/devices_planner_strategies/legacy.rb (hunk 1 above)
Comment (wfeldt) [reply src/lib/y2storage/proposal/devices_planner_strategies/legacy.rb]: Fine with me. But there's no consistency in the storage-ng code here.

File: src/lib/y2storage/proposal/devices_planner.rb (hunk 4 above)
Comment (wfeldt) [reply src/lib/y2storage/proposal/devices_planner.rb]: I'm not convinced. Having different errors for all kind of things provides a useless level of detail. In the end you'll just check for an error and not distinguish several cases while setting a strategy.

Is there anything bad with having a `DevicesPlanner::STRATEGIES` constant?

Comment (joseivanlopez) [reply src/lib/y2storage/proposal/devices_planner.rb]: Well, I am not against about using `STRATEGIES` or raise `ArgumentError`, but what I do not see correct is to raise always the same message error independently of the exception. For example, an exception could happen in `@settings.format` or in `STRATEGIES[strategy].new` even when the `STRATEGIES[strategy]` is known.

So, I agree with you propose, but maybe raising the exception only when the strategy is not known:

```
def initialize(settings, devicegraph)
  @settings = settings
  @devicegraph = devicegraph
  strategy = @settings.format
  if STRATEGIES[strategy]
    @strategy = STRATEGIES[strategy].new(settings, devicegraph)
  else
    err_msg = "Unsupported device planner strategy :#{strategy}"
    log.error err_msg
    raise ArgumentError, err_msg 
  end
end

```

**Comment by joseivanlopez** [comment]
Test coverage decreases a little bit, but it should be increased after next PRs related with this PBI.

**Pull request merged**